### ISSUE TITLE
Discussion of additional funktionalties and QC methods

### ISSUE THREAD

**Issue by maxmargraf** [OpenSenseAction/pypwsqc#23]
Here is a summary of functionality from intense-qc and its extension for sub-hourly data as a basis for discussion of what could be added to pypwsqc:

**1. intense-qc**
check individual gauges
* check if high percentiles (.99 and .95) of time rainfall time series are zero [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L236)
* return the largest n rainfall values [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L236)
* check how rainfall is distributed evenly over each day of the week [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L260)
* check how rainfall is didistributed over hours of day [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L275)
* searching for data with more than 5 data gaps longer than 2 consecutive steps are missing (on an hourly basis) [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L290)
* Pettitt check (change point test) [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L392)
* check if world records (from an internally used dataset) are broken [code}](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L411)
* flag too long dry or wet spells  (compared to the same internally dataset) [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L511)
* check if the amount of one time step is accumulated from more time steps before because of its intensity [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L616)
* check if high values are shown repeatedly which is unlikely [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L734)
* check if the minimal resolution changes over time [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L826)

check neighboring gauges
* get neighbors [code](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L859)
* compare [hourly](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L926), [daily](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L979) and [monthly](https://github.com/nclwater/intense-qc/blob/f189f4b386d73955dc41ed146cd055a797a6d605/intense/qc.py#L1062) neighors

**2. [SubHourlyQC](https://github.com/nclwater/SubHourlyQC ) to use on top of intense-qc** 
* checks if data is sub-hourly
* uses thresholds to reject unplausible rainfall
* various thresholds [are suggested](https://github.com/nclwater/SubHourlyQC/blob/b8ddcd6d8ce8f11f3b96e7b86b1763c442f3cbf5/SHQC_checks.py#L342) for different accumulation times (1min, 15min, 60min) and each month of the year

**Comment by lepetersson** [comment]
Summary of [titanlib](https://github.com/metno/titanlib/wiki/) functionalities. It emphases spatial checks and is written in C++ and has bindings for python and R. It was originally developed for temperature. 

Note that it deviates a bit from the checks presented in [this paper ](https://asr.copernicus.org/articles/17/153/2020/). Also note that [titan](https://github.com/metno/titan) and [titanlib ](https://github.com/metno/titanlib/) are different packages, I have not clarified how they are different.

strikethrough bullet points are rejected by Louise
--> refers to comments by Louise

- [Buddy check](https://github.com/metno/titanlib/wiki/Buddy-check) -compares an observation against its neighbours (i.e. buddies) and flags outliers over defined threshold --> redundant with FZ+HI filter in combination
- [Buddy event check](https://github.com/metno/titanlib/wiki/Buddy-event-check) - similar to the buddy check, except that observations are converted into yes/no values of exceeding a specified threshold. --> redundant with HI filter for high intensity events
- [Climatology range check](https://github.com/metno/titanlib/wiki/Climatology-range-check) - ~currently implemented only for temperature~
- [First guess test](https://github.com/metno/titanlib/wiki/First-guess-test) -implements a streamlined version of the Spatial Consistency Test (SCT, see below), which is less computationally intensive
- [Isolation check](https://github.com/metno/titanlib/wiki/Isolation-check) -  flags stations that have fewer than `num_min` buddies within a specified radius --> can be easily implemented and handy to identify isolated stations, for which most filters cannot be applied

**Comment by maxmargraf** [comment]
Some additional ideas came up during discussions at EGU with @JochenSeidel and Andras Bardossy regarding the indicator correlation filter in cases with no or sparse primary stations (reliable rain gauges).
* The PWS with the highest indicator correlation (and potentially no complaints from all other checks) can be used as primary stations. The number of PWS assumed as primary stations could be improved iteratively. 
* Vice versa from a PWS-only indicator correlation, the worst PWS can be removed in an iterative process until a certain quality is reached

**Comment by cchwala** [comment]
Yes, finding and removing duplicate locations should be made easy.

**Comment by cchwala** [comment]
@JochenSeidel can you post the pandas code that you used?

**Comment by cchwala** [comment]
Screenshot of the code, with comment based on our discussion that it should be done pair-wise which it probably is not doing with the code in the screenshot

<img width="527" alt="Bildschirmfoto 2024-05-14 um 11 02 14" src="https://github.com/OpenSenseAction/pypwsqc/assets/102827/067f3371-2f5a-4c06-9a2e-8f6d356e7708">

**Comment by cchwala** [comment]
Sounds good. 

>  This function should be split into identifying duplicate locations first and the choosing whether they should be kept/further used...

You could return the indices along the `id` dimension or the actual IDs which can then be used as index. The behavior should be as in [`pandas.DataFrame.duplicated`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) regarding the `keep` argument.

**Comment by JochenSeidel** [comment]
I just checked the `duplicate`function from `pandas`on a larger PWS data set (> 75.000 stations).
![grafik](https://github.com/OpenSenseAction/pypwsqc/assets/108326761/2196c47d-2e3a-40e3-9d53-098121c25c49)
The numbers are different indicating that some PWS have only one identical 'lat' or 'lon' value
If the operation is applied on both coordinate columns, fewer stations are found:
![grafik](https://github.com/OpenSenseAction/pypwsqc/assets/108326761/1628668e-7341-40b3-be4f-f891bbb17923)

**Comment by cchwala** [comment]
Ah, okay. So that means I was wrong with my concern about your usage of `.duplicated()`, right?

**Comment by JochenSeidel** [comment]
It looks like it does what it should do. I've created a small artificial test data set (attached) with some identical `lat` and/or `lon` values and it keeps the correct ones with unique coordinates. The single duplicate info for either `lat` or `lon` does not reveal much information though at it only counts the duplicate occurrences in the corresponding values. 
Before:
![grafik](https://github.com/OpenSenseAction/pypwsqc/assets/108326761/a81e9887-cbbb-4d38-8cd0-e5df3686b2b0)

After:
![grafik](https://github.com/OpenSenseAction/pypwsqc/assets/108326761/93799b66-eb43-41af-b4e5-eaf9b45b9c7f)

**Comment by cchwala** [comment]
👍 Very good

**Comment by JochenSeidel** [comment]
Attaching the file doesn't work, but I think we are safe to proceed with this.

**Comment by cchwala** [comment]
Should there be a function `get_duplicate_locations`, which would use pandas internally, or should we just list the approach using `to_dataframe()` and then `duplicated` in pandas?

`xarray` has a similar function [`drop_duplicates`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.drop_duplicates.html#xarray.DataArray.drop_duplicates) but no function to just get the indices of the duplicates like `DataFrame.duplicated`.

**Comment by cchwala** [comment]
Okay. Good to know. 

I suggest that you open a new issue for the final discussion and decision regarding functionality for getting duplicated coordinates. I am not yet sure if is is better to promote (in one of our example notebooks) a one-liner based on `.to_dataframe` and then using `pandas` or if we want to have one function. Let's discuss that in the new issue, which can then also be closed once things are done.

**Comment by JochenSeidel** [comment]
Final question: Should I open the new discussion in [poligrain](https://github.com/OpenSenseAction/poligrain) or here? As duplicates might also be relevant von CML I would go for poligrain

**Comment by cchwala** [comment]
yes, new issue in `poligrain` because, as you said, this will be a general functionality. please link to this issue then.

**Comment by cchwala** [comment]
I just found this recent AMT paper that also deals with QC of gauge and PWS data

"Adaptation of RainGaugeQC algorithms for quality control of rain gauge data from professional and non-professional measurement networks" https://doi.org/10.5194/amt-18-3229-2025

I have no time to check it, but there might be some info on effectiveness of different filters.
